File: rxkata/src/main/java/org/sergiiz/rxkata/CountriesService.java
package org.sergiiz.rxkata;

import java.util.List;
import java.util.Map;
import java.util.concurrent.FutureTask;

import io.reactivex.Observable;
import io.reactivex.Single;

interface CountriesService {

    Single<String> countryNameInCapitals(Country country);

    Single<Integer> countCountries(List<Country> countries);

    Observable<Long> listPopulationOfEachCountry(List<Country> countries);

    Observable<String> listNameOfEachCountry(List<Country> countries);

    Observable<Country> listOnly3rdAnd4thCountry(List<Country> countries);

    Single<Boolean> isAllCountriesPopulationMoreThanOneMillion(List<Country> countries);

    Observable<Country> listPopulationMoreThanOneMillion(List<Country> countries);

    /**
     * @param countriesFromNetwork an async task which is sometimes very very slow
     * @return the filtered values from the {@link FutureTask} or an {@link Observable#empty()} if there are no values within 1 second
     */
    Observable<Country> listPopulationMoreThanOneMillionWithTimeoutFallbackToEmpty(FutureTask<List<Country>> countriesFromNetwork);

    Observable<String> getCurrencyUsdIfNotFound(String countryName, List<Country> countries);

    Observable<Long> sumPopulationOfCountries(List<Country> countries);

    Single<Map<String, Long>> mapCountriesToNamePopulation(List<Country> countries);

}

File: src/main/java/Rx/CountriesService.kt
package Rx

import Rx.Country
import java.util.concurrent.FutureTask

import io.reactivex.Observable
import io.reactivex.Single

interface CountriesService {

    fun countryNameInCapitals(country: Country): Single<String>

    fun countCountries(countries: List<Country>): Single<Int>

    fun listPopulationOfEachCountry(countries: List<Country>): Observable<Long>

    fun listNameOfEachCountry(countries: List<Country>): Observable<String>
//
//    fun listOnly3rdAnd4thCountry(countries: List<Country>): Observable<Country>
//
//    fun isAllCountriesPopulationMoreThanOneMillion(countries: List<Country>): Single<Boolean>
//
//    fun listPopulationMoreThanOneMillion(countries: List<Country>): Observable<Country>

    /**
     * @param countriesFromNetwork an async task which is sometimes very very slow
     * @return the filtered values from the [FutureTask] or an [Observable.empty] if there are no values within 1 second
     */
//    fun listPopulationMoreThanOneMillionWithTimeoutFallbackToEmpty(countriesFromNetwork: FutureTask<List<Country>>): Observable<Country>
//
//    fun getCurrencyUsdIfNotFound(countryName: String, countries: List<Country>): Observable<String>
//
//    fun sumPopulationOfCountries(countries: List<Country>): Observable<Long>
//
//    fun sumPopulationOfCountries(countryObservable1: Observable<Country>,
//                                 countryObservable2: Observable<Country>): Observable<Long>
//
//    fun mapCountriesToNamePopulation(countries: List<Country>): Single<Map<String, Long>>
//
//    fun areEmittingSameSequences(countryObservable1: Observable<Country>,
//                                 countryObservable2: Observable<Country>): Single<Boolean>
}
